### PR TITLE
Bsp/spi

### LINKED ISSUE #122: [BSP] STM32f413 SPI Implementation
Adapt BPS's BSP_SPI code.

### PR DESCRIPTION
**Pull request by Cam0Cow**
Adapts BPS's SPI implementation. Closes #122

### CODE REVIEW

File: BSP/STM32F413/Src/BSP_SPI.C [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+
+#define SPI_PORT SPI3
+
+OS_SEM SPI_Update_Sem4;
+
+static void spi_post(void) {
+	OS_ERR err;
+	OSSemPost(&SPI_Update_Sem4, OS_OPT_POST_1, &err);
+	// TODO: error handling
```
Comment (chaseblock) [inline BSP/STM32F413/Src/BSP_SPI.C:15]: Ideally, this should be addressed before merging this. If you want to wait on this, at least make a plan or issue to address this.

File: BSP/Inc/BSP_SPI.h [hunk 2]
```diff
@@ -38,6 +39,9 @@
 #define SPI_GPIOB       0x19
 #define SPI_OLATB       0x1A
 
+// For synchronizing SPI
+extern OS_SEM SPI_Update_Sem4;
```
Comment (chaseblock) [inline BSP/Inc/BSP_SPI.h:43]: Why does this need to be declared in a visible way? Is something other than BSP_SPI.c going to use it?

File: BSP/STM32F413/Src/BSP_SPI.C [hunk 3]
```diff
@@ -0,0 +1,162 @@
+/* Copyright (c) 2020 UT Longhorn Racing Solar */
```
Comment (chaseblock) [inline BSP/STM32F413/Src/BSP_SPI.C:1]: Make the name of this file `BSP_SPI.c` (lowercase c)

Comment (chaseblock) [reply BSP/STM32F413/Src/BSP_SPI.C:1]: As a side note, `BSP_CAN.c` also needs to be corrected.

File: BSP/STM32F413/Src/BSP_SPI.C [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+ * @param   rxBuf   data array to store the data that is received.
+ * @param   rxLen   length of data array.
+ * @return  None
+ */
+void BSP_SPI_Read(uint8_t *rxBuf, uint8_t rxLen) {
+    for(uint32_t i = 0; i < rxLen; i++){
+		rxBuf[i] = SPI_WriteRead(0x00);
+	}
+
+}
```
Comment (chaseblock) [inline BSP/STM32F413/Src/BSP_SPI.C:142]: Note that these should not be called from an ISR. Maybe good to include a warning about that in the comment header.

File: BSP/STM32F413/Src/BSP_UART.c [hunk 5]
```diff
@@ -202,7 +202,7 @@ void BSP_UART_Init(UART_t uartSelect) {
  * @param   str pointer to buffer string
  * @return  number of bytes that was read
  */
-uint32_t BSP_UART_ReadLine(UART_t uartSelect, char *str) {
+uint32_t BSP_UART_Read(UART_t uartSelect, char *str) {
```
Comment (chaseblock) [inline BSP/STM32F413/Src/BSP_UART.c:205]: I'm curious: why the name change?

File: BSP/Inc/BSP_SPI.h (hunk 2 above)
Comment (Cam0Cow) [reply BSP/Inc/BSP_SPI.h:43]: I just wanted to keep it in the same place as the rest of the semaphores, but reducing visibility sounds like a good idea.

File: BSP/STM32F413/Src/BSP_UART.c (hunk 5 above)
Comment (Cam0Cow) [reply BSP/STM32F413/Src/BSP_UART.c:205]: We wrote the simulator with that function name, so it seemed silly to go back and change all instances of it to BPS's terminology.

File: BSP/STM32F413/Src/BSP_SPI.C (hunk 3 above)
Comment (JimothyGreene) [reply BSP/STM32F413/Src/BSP_SPI.C:1]: Still should change this

File: BSP/STM32F413/Src/BSP_SPI.C [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+	GPIO_InitStruct.GPIO_Pin = GPIO_Pin_5 | GPIO_Pin_6 | GPIO_Pin_7;
+	GPIO_InitStruct.GPIO_Mode = GPIO_Mode_AF;
+	GPIO_InitStruct.GPIO_PuPd = GPIO_PuPd_UP;
+	GPIO_InitStruct.GPIO_Speed = GPIO_Speed_50MHz;
+	GPIO_InitStruct.GPIO_OType = GPIO_OType_PP;
+	GPIO_Init(GPIOA, &GPIO_InitStruct);
+
+	GPIO_PinAFConfig(GPIOC, GPIO_PinSource5, GPIO_AF_SPI3);
+	GPIO_PinAFConfig(GPIOC, GPIO_PinSource6, GPIO_AF_SPI3);
+	GPIO_PinAFConfig(GPIOC, GPIO_PinSource7, GPIO_AF_SPI3);
```
Comment (JimothyGreene) [inline BSP/STM32F413/Src/BSP_SPI.C:82]: I think these should be `GPIOA` for Controls

File: BSP/STM32F413/Src/BSP_SPI.C [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+/** SPI_WriteRead
+ * @brief   Sends and receives a byte of data on the SPI line.
+ * @param   txData single byte that will be sent to the device.
+ * @return  rxData single byte that was read from the device.
+ */
+static uint8_t SPI_WriteRead(uint8_t txData){
+
+	SPI_TypeDef *bus = SPI_PORT;
+	
+	SPI_Wait(bus);
```
Comment (JimothyGreene) [inline BSP/STM32F413/Src/BSP_SPI.C:39]: Is this supposed to take a parameter? From the function definition above it seems like it doesn't

**Review by JimothyGreene — CHANGES_REQUESTED** [review]

File: BSP/STM32F413/Src/BSP_SPI.C (hunk 3 above)
Comment (Cam0Cow) [reply BSP/STM32F413/Src/BSP_SPI.C:1]: For some reason I can't change the filename. It might be a problem with my WSL configuration.

Comment (JimothyGreene) [reply BSP/STM32F413/Src/BSP_SPI.C:1]: Yeah it's a weird issue. The way I got around it (probably not the proper way) was to delete the file, make a commit, then create a new file with the lowercase 'c' and put all the code in that and commit it again. It's definitely not the best practice but it worked

**Review by JimothyGreene — COMMENTED** [review]
Just try and fix that capitalization in the filename. But other than that, it all seems good. All this BSP stuff is approved with a note that we need to test obviously. But it seems fine from reading it.

**Review by JimothyGreene — APPROVED** [review]

**Pull request merged**